proxy_agent/src/redirector/windows/bpf_api.rs

```diff
@@ -4,43 +4,83 @@
 #![allow(non_snake_case)]
 
 use super::bpf_obj::*;
+use crate::common::logger;
 use libloading::{Library, Symbol};
 use once_cell::sync::Lazy;
 use proxy_agent_shared::misc_helpers;
 use std::ffi::{c_char, c_int, c_uint, c_void, CString};
 use std::io::{Error, ErrorKind};
 use std::path::PathBuf;
+use std::{env, thread};
 
-static mut EBPF_API: Lazy<Option<Library>> = Lazy::new(|| None);
+static EBPF_API: Lazy<Option<Library>> = Lazy::new(init_ebpf_lib);
+const EBPF_API_FILE_NAME: &str = "EbpfApi.dll";
 
-pub fn load_ebpf_api(bpf_api_file_path: PathBuf) -> std::io::Result<()> {
-    unsafe {
-        match Library::new(bpf_api_file_path.as_path()) {
-            Ok(api) => {
-                *EBPF_API = Some(api);
+pub fn ebpf_api_is_loaded() -> bool {
+    EBPF_API.is_some()
+}
+
+fn init_ebpf_lib() -> Option<Library> {
+    for _ in 0..5 {
+        let program_files_dir = env::var("ProgramFiles").unwrap_or("C:\\Program Files".to_string());
+        let program_files_dir = PathBuf::from(program_files_dir);
+        let ebpf_for_windows_dir = program_files_dir.join("ebpf-for-windows");
+        let bpf_api_file_path = ebpf_for_windows_dir.join(EBPF_API_FILE_NAME);
+        logger::write_information(format!(
+            "Try to load ebpf api file from: {}",
+            misc_helpers::path_to_string(bpf_api_file_path.to_path_buf())
+        ));
+        match load_ebpf_api(bpf_api_file_path) {
+            Ok(ebpf_lib) => {
+                return Some(ebpf_lib);
+            }
+            Err(e) => {
+                logger::write_warning(format!("{}", e));
+            }
+        }
+
+        logger::write_warning("Try to load ebpf api file from default system path".to_string());
+        match load_ebpf_api(PathBuf::from(EBPF_API_FILE_NAME)) {
+            Ok(ebpf_lib) => {
+                return Some(ebpf_lib);
             }
             Err(e) => {
-                let message = format!(
-                    "Loading ebpf api from file {} failed with error: {}",
-                    misc_helpers::path_to_string(bpf_api_file_path.to_path_buf()),
-                    e
-                );
-                return Err(Error::new(ErrorKind::Other, message));
+                logger::write_warning(format!("{}", e));
             }
         }
+
+        thread::sleep(std::time::Duration::from_millis(10));
     }
 
-    Ok(())
+    // after 5 tries, still can't load ebpf api, return None
+    None
+}
+
+fn load_ebpf_api(bpf_api_file_path: PathBuf) -> std::io::Result<Library> {
+    // load ebpf api library dynamically
+    // For thread local init / cleanup -- it only stores and cleans up the device handle to ebpf core.
+    //      So it should be transparent to the caller, and does not impose any extra conditions.
+    // Considering DllMain (for process attach / detach) will be called only once,
+    //      there should not be thread safety concerns too.
```

Comment:
Technically, `Library::new` is unsafe when called with a library name and the library search path is modified at the same time. You're not using it that way in this module, but the way you've written it makes it easy to refactor code elsewhere and break the assumption here that the function is never being called in that way.

Since you're hard-coding the library path in the function calling this function, I'd recommend moving that down into this function since it's important to the safety of this function. The comment can then be something like:

```
// Safety: Loading a library on  Windows has the following safety requirements:
// 1. The safety requirements of the library initialization and/or termination routines must be met.
// The eBPF for Windows library does not have any safety requirements for its initialization and de-initialization function[0].
// [0] https://github.com/microsoft/ebpf-for-windows/blob/Release-v0.17.1/ebpfapi/dllmain.cpp
//
// 2. Calling this function is not safe in multi-thread scenarios where a library file name is provided and the library
//     search path is modified.
// We satisfy the this requirement by providing the absolute path to the library.
```

Reply:
Thanks for your words, I will update to the comments.